### DISCUSSION TITLE
Can/Should twig components replace embeds/includes/macros?

### DISCUSSION THREAD

**Discussion by tacman** [symfony/symfony#51755] [Q&A]
I *love* the Symfony UX components.  It feels very natural to move complex logic to PHP and let twig do the rendering.

Is there any reason to not replace includes, embeds and macros with twig components?  And move twig logic into PHP via a component?

In particular, I struggle with the KnpMenuBundle templates, and I'd like to move some of the twig logic to PHP.  For example, the 'item' block in https://github.com/KnpLabs/KnpMenu/blob/master/src/Knp/Menu/Resources/views/knp_menu.html.twig#L51-L95

```twig
{% block item %}
{% if item.displayed %}
{# building the class of the item #}
    {%- set classes = item.attribute('class') is not empty ? [item.attribute('class')] : [] %}
    {%- if matcher.isCurrent(item) %}
        {%- set classes = classes|merge([options.currentClass]) %}
    {%- elseif matcher.isAncestor(item, options.matchingDepth) %}
        {%- set classes = classes|merge([options.ancestorClass]) %}
    {%- endif %}
    {%- if item.actsLikeFirst %}
        {%- set classes = classes|merge([options.firstClass]) %}
    {%- endif %}
    {%- if item.actsLikeLast %}
        {%- set classes = classes|merge([options.lastClass]) %}
    {%- endif %}


    {# Mark item as "leaf" (no children) or as "branch" (has children that are displayed) #}
    {% if item.hasChildren and options.depth is not same as(0) %}
        {% if options.branch_class is not empty and item.displayChildren %}
            {%- set classes = classes|merge([options.branch_class]) %}
        {% endif %}
    {% elseif options.leaf_class is not empty %}
        {%- set classes = classes|merge([options.leaf_class]) %}
    {%- endif %}


    {%- set attributes = item.attributes %}
    {%- if classes is not empty %}
        {%- set attributes = attributes|merge({'class': classes|join(' ')}) %}
    {%- endif %}
{# displaying the item #}
    {% import _self as knp_menu %}
    <li{{ knp_menu.attributes(attributes) }}>
        {%- if item.uri is not empty and (not matcher.isCurrent(item) or options.currentAsLink) %}
        {{ block('linkElement') }}
        {%- else %}
        {{ block('spanElement') }}
        {%- endif %}
{# render the list of children#}
        {%- set childrenClasses = item.childrenAttribute('class') is not empty ? [item.childrenAttribute('class')] : [] %}
        {%- set childrenClasses = childrenClasses|merge(['menu_level_' ~ item.level]) %}
        {%- set listAttributes = item.childrenAttributes|merge({'class': childrenClasses|join(' ') }) %}
        {{ block('list') }}
    </li>
{% endif %}
{% endblock %}
```

There is very little rendering here, it's almost all logic, merging properties (and the twig syntax is somewhat verbose to do so).  So I'm wondering if it's worth diving into this code and replacing the logic with a component, and then

```twig
<twig:block name="item"> 
   <twig:menuItem :item="item" :options="options" />
</twig:block>
```

and creating the menuItem component where the PHP handles the attributes, then passes it off to a much simpler twig template that renders the list item (or div, which is indeed part of the refactoring I'm trying to do.)

@kbond or @weaverryan, any advice?  Am I so enamored with ux components that I'm thinking about using them where a macro or include or keeping the logic in twig would be more appropriate?

**Comment by weaverryan** [comment]
I think there is a lot of potential to do things like this - I would also rather render a Twig component :). But it's not about needing to replace embeds/includes/macros... I think we could rather propose a Twig component into libraries like KnpMenu, or have a central place where you can grab a component (PHP class + template) for a particular library (e.g. KnpMenu) so you can quickly use it in your app. I've had similar thoughts about rendering forms - https://twitter.com/weaverryan/status/1706342492612841669 - but we don't need to replace the core form rendering system, we just need to make a new set of components available for people to use.

**Reply by tacman** [reply]
Makes sense.

In the specific case of KnpMenu, most of the PHP logic is already in things like ListRenderer.php, using the format() method to actually render the HTML.  It's still a matter of figuring out where and how to leverage that and Twig.

And by "replace" I didn't mean replace it in Symfony, but rather in my own code and bundles.  In particular, macros seem outdated in light of components.

**Reply by weaverryan** [reply]
I definitely use TwigComponents now often instead of other things. For pre-made render functions like KnpMenu (where the work is already done for you), implementing a Twig component instead takes time, but gives you more flexibility. As a community, if we had some pre-made Twig components for these common problems, that would alleviate that. Best of both worlds.

**Comment by tacman** [comment]
On a related note, is there a way to register anonymous components in a bundle?  Continuing with the KnpMenu example, is there a way to create /templates/components/menu/root.html.twig and be able to call <twig:menu:root :item="item" />  

Currently I registered all the components that don't inject services from the Bundle.php configuration:

```php
        foreach (
            [AlertComponent::class,
                     AccordionComponent::class,
                     CarouselComponent::class,
                     DropdownComponent::class,
                     DividerComponent::class,
                     CardComponent::class,
                     ButtonComponent::class,
                     BadgeComponent::class,
                     LinkComponent::class
                 ] as $componentClass
        ) {
            $builder->register($componentClass)->setAutowired(true)->setAutoconfigured(true);
        }
```

Is there a way to add a bundle to whatever scans the templates/components directory?

**Reply by kbond** [reply]
> is there a way to register anonymous components in a bundle?

[This PR](https://github.com/symfony/ux/pull/1140) I think could help but I don't think bundle's are taken into account.
